driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -130,41 +120,56 @@ public boolean isClosed() {
         return closed;
     }
 
-    ServerCursor getServerCursor() {
-        return cursor;
+    private boolean getAndSetClosed() {
+        boolean oldClosed;
+        synchronized (lock) {
+            oldClosed = closed;
+            closed = true;
+        }
+        return oldClosed;
+    }
+
+    private ServerCursor getAndSetCursor(final ServerCursor cursor) {
+        ServerCursor oldCursor;
+        synchronized (lock) {
+            oldCursor = this.cursor;
+            this.cursor = cursor;
+        }
+        return oldCursor;
     }
 
     private boolean limitReached() {
         return Math.abs(limit) != 0 && count >= Math.abs(limit);
     }
 
-    private void getMore(final SingleResultCallback<List<T>> callback) {
+    private void getMore(final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
         connectionSource.getConnection(new SingleResultCallback<AsyncConnection>() {
```

Comment:
> The most recent commit has a similar race condition.
> 
> T1: next() sets cursor to null, and gets a non-null cursor
> T2: close() sets cursor to null, and gets a null cursor
> T3: close() (synchronously) releases connectionSource since cursor is null
> T4: next() calls connectionSource.getConnection, but refcount is 0

T4  would throw an error as its called after close(), but you're right we don't close the final getMore cursor.

I've added a new patch and some more unit tests.